### PR TITLE
Write test for app/controllers/webui/project_controller.rb#remove_maintained_project

### PR DESCRIPTION
**Pull request by Ana06**
Write test for `app/controllers/webui/project_controller.rb#remove_maintained_project` and for some functions used in the before filter for `remove_maintained_project`.

I also modified `project_controller#remove_maintained_project` function by rendering an error when the MaintainedProject is not found and by using `destroy` in the if condition instead of `save` after doing `delete`.

### CODE REVIEW

**Comment by bgeuken** [comment]
Please add a [webui] tag to the first commit:-)

File: src/api/app/controllers/webui/project_controller.rb [hunk 1]
```diff
@@ -709,8 +709,8 @@ def add_maintained_project
 
   def remove_maintained_project
     authorize @project, :update?
-    @project.maintained_projects.delete(MaintainedProject.find_by(project: @maintained_project))
-    if @project.save
+    maintained_project = MaintainedProject.find_by(project: @maintained_project)
+    if maintained_project && @project.maintained_projects.destroy(maintained_project)
```
Comment (bgeuken) [inline src/api/app/controllers/webui/project_controller.rb]: Let's move ```@project.maintained_projects.destroy(maintained_project)``` inside the if block. Like "if it exists then delete it, store the project, etc, else ..."

Comment (bgeuken) [inline src/api/app/controllers/webui/project_controller.rb]: Unless destroy gives any return value that would indicate if deleting it succeeded or failed.

**Comment by mdeniz** [comment]
@openSUSE/open-build-service, generally speaking, not tied to this PR only: 

should we add the tests first, and then refactor instead of the other way around? and should we use 2 PRs for that?

I mean, if you change the method and create tests for it at the same time then you don't know if you have refactored it right.. maybe you change the workflow and then you created the tests in consequence, and this is wrong.

So my opinion is, create 2 PRs:
 - First create the tests that covers the action
 - Second, refactor the action and see that the tests are not failing, so you didn't change the workflow

What do you thinkg?

**Comment by bgeuken** [comment]
Preferably tests should be written first, or it should be verified that there are some already.

For smaller refactorings I wouldn't make it mandatory to have seperate PRs though. The developer should just make sure that he creates and runs the test before and after the refactoring.

It's really cool when people do refactorings and not just do the minimal stuff necessary to finish a card. So I would try to minimize the obstacles.

For larger refactorings / more complicated changes a seperate PR is a good idea.

File: src/api/spec/controllers/webui/project_controller_spec.rb [hunk 2]
```diff
@@ -715,4 +715,45 @@
...
+    context "#remove_maintained_project succeeds with maintenance kind for a valid maintained project" do
+      before do
+        user.home_project.update(kind: 'maintenance')
+        post :remove_maintained_project, project: user.home_project, maintained_project: maintained_project.project.name
+      end
+
+      it { expect(flash[:notice]).to eq("Removed #{maintained_project.project.name} from maintenance") }
+      it { is_expected.to redirect_to(action: 'maintained_projects', project: user.home_project) }
+    end
+  end
```
Comment (bgeuken) [inline src/api/spec/controllers/webui/project_controller_spec.rb]: @Ana06 @ChrisBr I don't think we should have a separate controller test for before filters. Maybe in some rare edge cases like authentification related before filters. But not here.

This seems to be implementation details of how we handle the corresponding http request. So I think it would be better to integrate these tests into the other "POST #remove_maintained_project" part.
And if it turns out that we do not catch the ActiveRecord::RecordNotFound and show a proper error message, we should fix this:-)

File: src/api/spec/controllers/webui/project_controller_spec.rb [hunk 3]
```diff
@@ -715,4 +715,45 @@
...
+    before do
+      login user
+      user.home_project.update(kind: 'maintenance')
+    end
+
+    it "#remove_maintained_project raise excepction with invalid maintained project" do
+      expected = expect do
+        post :remove_maintained_project, project: user.home_project, maintained_project: "invalid"
+      end
+      expected.to raise_exception ActiveRecord::RecordNotFound
```
Comment (bgeuken) [inline src/api/spec/controllers/webui/project_controller_spec.rb]: Didn't try it yet, but something like this might also work:
```
post remove_maintained_project, project: user.home_project, maintained_project: "invalid"
expect(response).to raise_error(ActiveRecord::RecordNotFound)
```

Comment (Ana06) [inline src/api/spec/controllers/webui/project_controller_spec.rb]: It doesn't work. As it is said in the [raise_error examples](http://www.relishapp.com/rspec/rspec-expectations/v/3-2/docs/built-in-matchers/raise-error-matcher):

> raise_error matcher specifies that a block of code raises an error

So without the block the `post` fails before reaching the expect statement.

Comment (bgeuken) [inline src/api/spec/controllers/webui/project_controller_spec.rb]: I see. What about this one:
```
expect {
  post :remove_maintained_project, project: user.home_project, maintained_project: "invalid"
}.to raise_exception ActiveRecord::RecordNotFound
``` ?

@openSUSE/open-build-service opinions? I slightly prefer the compact version.

File: src/api/app/controllers/webui/project_controller.rb (hunk 1 above)
Comment (Ana06) [inline src/api/app/controllers/webui/project_controller.rb]: `destroy` returns the object that was destroyed (which acts as `true`) if it succeeded and `false` otherwise. 

> There's a series of callbacks associated with destroy. If the `before_destroy` callback throws `:abort` the action is cancelled and destroy returns `false`. 

http://edgeapi.rubyonrails.org/classes/ActiveRecord/Persistence.html#method-i-destroy

File: src/api/spec/controllers/webui/project_controller_spec.rb (hunk 3 above)
Comment (mdeniz) [inline src/api/spec/controllers/webui/project_controller_spec.rb]: Me too , I prefer using the block directly instead of using a local variable for it

Comment (Ana06) [inline src/api/spec/controllers/webui/project_controller_spec.rb]: ok, I'll change it! :wink:

File: src/api/app/controllers/webui/project_controller.rb (hunk 1 above)
Comment (bgeuken) [inline src/api/app/controllers/webui/project_controller.rb]: Okay.Then nevermind me:-)

File: src/api/spec/controllers/webui/project_controller_spec.rb (hunk 2 above)
Comment (ChrisBr) [inline src/api/spec/controllers/webui/project_controller_spec.rb]: Okay, agreed!

**Comment by ChrisBr** [comment]
@Ana06 if you fix @bgeuken suggestions it LGTM :+1:

**Comment by ChrisBr** [comment]
Ah an CI is still failing
```
rspec ./spec/controllers/webui/project_controller_spec.rb:792 
rspec ./spec/controllers/webui/project_controller_spec.rb:802 
rspec ./spec/controllers/webui/project_controller_spec.rb:801 
```

**Comment by bgeuken** [comment]
@Ana06 You might have to recreate the VCR cassettes. Just delete them (the ones your PR creates) and re-create them by running the tests. Otherwise we end up with a lot of additional VCR records that are not used.

**Comment by Ana06** [comment]
@mdeniz I agree with you that we should add test first and then refactor but the problem here was that it was not possible that the `.save` (https://github.com/openSUSE/open-build-service/blob/master/src/api/app/controllers/webui/project_controller.rb#L702) fails, so it was not possible to test the fail case. And also I didn't like getting an exception in case that the project doesn't exist. So, I was not refactoring really as I wanted to change the functionality of the function a little bit.  :persevere:

**Comment by bgeuken** [comment]
Let's move the discussion about refactorings to the mailing list. Since it's a more general question.
